test 12

leviychuvak opened the pull request.

leviychuvak committed:
add files of test 12

BlindBlackHole commented on main.cpp:
```diff
@@ -0,0 +1,37 @@
+#include "Circle.h"
+#include "Triangle.h"
+
+double getLargestRadius(std::vector<Shape*> v){
```
const&

BlindBlackHole commented on main.cpp:
```diff
@@ -0,0 +1,37 @@
+double getLargestRadius(std::vector<Shape*> v){
+    double largestRadius = 0.0;
+
+    for (int i = 0; i < v.size(); ++i) {
```
поскольку ты никак не используешь индекс (кроме получения элемента), лучше использовать range-based for цикл

BlindBlackHole replied:
о, и подумай как этот алгоритм можно написать через [std::max_element](https://en.cppreference.com/w/cpp/algorithm/max_element)
Возможно будет меньше кода или он будет аккуратней

BlindBlackHole commented on main.cpp:
```diff
@@ -0,0 +1,37 @@
+double getLargestRadius(std::vector<Shape*> v){
+    double largestRadius = 0.0;
+
+    for (int i = 0; i < v.size(); ++i) {
```
и кст, для обращения к элементам вектор юзает индекс типа size_t, а не int. Поэтому чтоб не было неявных кастов на каждой итерации, лучше юзать его

BlindBlackHole commented on main.cpp:
```diff
@@ -0,0 +1,37 @@
+    v.push_back(new Circle(Point(4, 5, 6), 13));
+    v.push_back(new Circle(Point(4, 5, 6), 3));
+
+    for (int i = 0; i < v.size(); ++i) {
```
range-based for

BlindBlackHole commented on main.cpp:
```diff
@@ -0,0 +1,37 @@
+
+    std::cout << "\nThe largest radius is: " << getLargestRadius(v) << '\n'; // реализуйте эту функцию
+
+    for (int i = 0; i < v.size(); ++i) {
```
range-based for

BlindBlackHole commented on main.cpp:
```diff
@@ -0,0 +1,37 @@
+        std::cout<<*v[i]<<std::endl;
+    }
+
+    std::cout << "\nThe largest radius is: " << getLargestRadius(v) << '\n'; // реализуйте эту функцию
```
коммент можно удалить

BlindBlackHole commented on Triangle.h:
```diff
@@ -0,0 +1,24 @@
+class Triangle: public Shape
+{
+public:
+    explicit Triangle(const Point &aFirst = Point(0,0,0),
```
можно не писать const Point &aFirst = Point(0,0,0), а const Point &aFirst = {} or const Point &aFirst = Point()

BlindBlackHole commented on Triangle.cpp:
```diff
@@ -0,0 +1,10 @@
+#include "Triangle.h"
+
+Triangle::Triangle(const Point &aFirst, const Point &aSecond, const Point &aThird)
+    : firstP(aFirst), secondP(aSecond), thirdP(aThird)
```
хз, будет ли написано на сайте об этом, поэтому скажу сейчас:
Если внутри класса ты хранишь значения, а не ссылки, то считается хорошей практикой принимать в конструкторы/сеттеры копию этих перемнных и мувать их внутрь класса.

BlindBlackHole replied:
кст в данном случае это тебя избавит от лишнего копирования. Потому что ты принимаешь временные объекты. Ну как пройдешь мув семантику ты все поймешь, поэтому сейчас не страшно

BlindBlackHole commented on Point.h:
```diff
@@ -0,0 +1,22 @@
+        :x(aX), y(aY), z(aZ)
+    {}
+
+    friend std::ostream& operator<<(std::ostream &out, const Point &p);
```
я бы лучше писал так Тут не обязательно делать friend 
std::ostream& operator<<(std::ostream &out) const;

BlindBlackHole replied:
в других местах тоже исправь

leviychuvak committed:
replaced for i to range-based for

leviychuvak committed:
Use std::move() in constructors
Use constructors without parameters instead of constructor with default parameters

leviychuvak committed:
delete friend

BlindBlackHole commented on main.cpp:
```diff
@@ -0,0 +1,37 @@
+    v.push_back(new Circle(Point(4, 5, 6), 13));
+    v.push_back(new Circle(Point(4, 5, 6), 3));
+
+    for (auto & i : v) {
```
не нужно &

BlindBlackHole commented on main.cpp:
```diff
@@ -0,0 +1,37 @@
+
+    std::cout << "\nThe largest radius is: " << getLargestRadius(v) << '\n';
+
+    for (auto & i : v) {
```
тоже не нужно &

BlindBlackHole commented on main.cpp:
```diff
@@ -0,0 +1,37 @@
+    double largestRadius = 0.0;
+
+    for(const auto& shape : v){
+        if (auto* temp = dynamic_cast<Circle*>(shape)){
```
раньше лучше было. Вложенные if такое себе

BlindBlackHole commented on Circle.cpp:
```diff
@@ -0,0 +1,14 @@
+#include "Circle.h"
+
+Circle::Circle(const Point aCenter, double aRadius)
```
const не нужно , потом поймешь почему

leviychuvak committed:
delete const from constructors
change range-based for

leviychuvak committed:
change realization of getLargestRadius() with using std::max_element
